Source/WebCore/rendering/style/PositionArea.h

```diff
@@ -25,43 +25,76 @@
  */
 
 #pragma once
+#include "BoxSides.h"
+#include "WritingMode.h"
 
 namespace WebCore {
 
 // The axis that the span specifies.
+// Represented as 3 bits, with the last one indicating the corresponding [Logical]BoxAxis value.
 enum class PositionAreaAxis : uint8_t {
-    // Physical axes. Implies self is PositionAreaSelf::No, as physical axes
-    // do not depend on writing mode of any elements.
-    Horizontal,
-    Vertical,
-
-    // Logical axes.
-    X,
-    Y,
-    Block,
-    Inline,
-};
+    // Physical axes × Physical directions.
+    Horizontal = 0b000,
+    Vertical   = 0b001,
 
-// Specifies which tile(s) on the axis that the position-area span occupies.
-enum class PositionAreaTrack : uint8_t {
-    // First tile.
-    Start,
+    // Physical axes × Logical directions.
+    X = 0b010,
+    Y = 0b011,
 
-    // First and center tiles.
-    SpanStart,
+    // Logical axes × Logical directions.
+    Inline = 0b110,
+    Block  = 0b111,
+};
 
-    // Last tile.
-    End,
+static inline bool isPositionAreaAxisLogical(const PositionAreaAxis positionAxis)
+{
+    return static_cast<uint8_t>(positionAxis) & 0b100;
+}
+
+static inline bool isPositionAreaDirectionLogical(const PositionAreaAxis positionAxis)
+{
+    return static_cast<uint8_t>(positionAxis) & 0b010;
+}
+
+static inline BoxAxis mapPositionAreaAxisToPhysicalAxis(const PositionAreaAxis positionAxis, const WritingMode writingMode)
+{
+    auto physicalAxis = static_cast<uint8_t>(positionAxis) & 0b001;
+    if (isPositionAreaAxisLogical(positionAxis) && writingMode.isVertical())
+        physicalAxis = !physicalAxis;
+    return static_cast<BoxAxis>(physicalAxis);
+}
+
+static inline LogicalBoxAxis mapPositionAreaAxisToLogicalAxis(const PositionAreaAxis positionAxis, const WritingMode writingMode)
+{
+    auto logicalAxis = static_cast<uint8_t>(positionAxis) & 0b001;
+    if (!isPositionAreaAxisLogical(positionAxis) && writingMode.isVertical())
+        logicalAxis = !logicalAxis;
+    return static_cast<LogicalBoxAxis>(logicalAxis);
+}
+
+// Specifies which tracks(s) on the axis that the position-area span occupies.
+// Represented as 3 bits: start track, center track, end track.
+enum class PositionAreaTrack : uint8_t {
```

Comment:
Have you considered using a `WTF::OptionSet` instead of doing manual bit handling?

Reply:
This should look, to the outside, like an enum. It's only internal methods like `invertPositionAreaTrack` that need to know about its internal information hacks.